services/dynamic-scheduler/src/simcore_service_dynamic_scheduler/services/generic_scheduler/_store.py

```diff
@@ -0,0 +1,432 @@
+
+
+def _get_scheduler_data_hash_key(*, schedule_id: ScheduleId) -> str:
+    # SCHEDULE_NAMESPACE:SCHEDULE_ID
+    # - SCHEDULE_NAMESPACE: namespace prefix
+    # - SCHEDULE_ID: the unique scheudle_id assigned
+    # Example:
+    # - SCH:00000000-0000-0000-0000-000000000000
+    return f"{_SCHEDULE_NAMESPACE}:{schedule_id}"
+
+
+def _get_step_hash_key(
+    *,
+    schedule_id: ScheduleId,
+    operation_name: OperationName,
+    group_name: StepGroupName,
+    step_name: StepName,
+    is_creating: bool,
+) -> str:
+    # SCHEDULE_NAMESPACE:SCHEDULE_ID:STEPS:OPERATION_NAME:GROUP_SHORT_NAME:STEP_NAME:IS_CREATING
+    # - SCHEDULE_NAMESPACE: namespace prefix
+    # - SCHEDULE_ID: the unique scheudle_id assigned
+    # - CONSTANT: the constant "STEPS"
+    # - OPERATION_NAME form the vairble's name during registration
+    # - GROUP_SHORT_NAME
+    #   -> "{index}(S|P)[R]": S=single or P=parallel and optinally, "R" if steps should be repeated forever
+    # - IS_CREATING: "C" (create) or "R" (revert)
+    # - STEP_NAME form it's class
+    # Example:
+    # - SCH:00000000-0000-0000-0000-000000000000:STEPS:START_SERVICE:0S:C:BS1
+    is_creating_str = _get_is_creating_str(is_creating=is_creating)
+    return f"{_SCHEDULE_NAMESPACE}:{schedule_id}:{_STEPS_KEY}:{operation_name}:{group_name}:{is_creating_str}:{step_name}"
+
+
+def _get_group_hash_key(
+    *,
+    schedule_id: ScheduleId,
+    operation_name: OperationName,
+    group_name: StepGroupName,
+    is_creating: bool,
+) -> str:
+    # SCHEDULE_NAMESPACE:SCHEDULE_ID:GROUPS:OPERATION_NAME:GROUP_SHORT_NAME:IS_CREATING
+    # - SCHEDULE_NAMESPACE: namespace prefix
+    # - SCHEDULE_ID: the unique scheudle_id assigned
+    # - CONSTANT: the constant "GROUPS"
+    # - OPERATION_NAME form the vairble's name during registration
+    # - GROUP_SHORT_NAME
+    #   -> "{index}(S|P)[R]": S=single or P=parallel and optinally, "R" if steps should be repeated forever
+    # - IS_CREATING: "C" (create) or "R" (revert)
+    # Example:
+    # - SCH:00000000-0000-0000-0000-000000000000:GROUPS:START_SERVICE:0S:C
+    is_creating_str = _get_is_creating_str(is_creating=is_creating)
+    return f"{_SCHEDULE_NAMESPACE}:{schedule_id}:{_GROUPS_KEY}:{operation_name}:{group_name}:{is_creating_str}"
+
+
+def _get_operation_context_hash_key(
+    *, schedule_id: ScheduleId, operation_name: OperationName
+) -> str:
+    # SCHEDULE_NAMESPACE:SCHEDULE_ID:STEPS:OPERATION_NAME
+    # - SCHEDULE_NAMESPACE: namespace prefix
+    # - SCHEDULE_ID: the unique scheudle_id assigned
+    # - CONSTANT: the constant "OP_CTX"
+    # - OPERATION_NAME form the vairble's name during registration
+    # Example:
+    # - SCH:00000000-0000-0000-0000-000000000000:OP_CTX:START_SERVICE
+    return (
+        f"{_SCHEDULE_NAMESPACE}:{schedule_id}:{_OPERATION_CONTEXT_KEY}:{operation_name}"
+    )
+
+
+def _dumps(obj: Any) -> str:
+    # NOTE: does not support `sets` and `tuples` they get serialised to lists
+    return json_dumps(obj)
```

Comment:
Q: what is the purpose of this extra indirection? you want in the future to support other type of serialization that is not `json` formatted?